webextension/privileged/ui/CloudDownloadsView.jsm

```diff
@@ -0,0 +1,738 @@
+        return;
+      }
+
+      const downloadType = downloadElement._shell.element.getAttribute("cloudstorage");
+      // Downloaded Item is saved in local download folder
+      if (downloadType === "local") {
+        if (this.providers.size > 1) {
+          let subMenuItem =  this._setMoveDownloadMenuPopUpAttributes(menuItem);
+          this.providers.forEach((value, key) => {
+            subMenuItem = this._setMenuItemAttributes(subMenuItem, key, value, downloadElement);
+            subMenuItem = subMenuItem.nextSibling;
+          });
+        } else {
+          // Display available provider as context menu option
+          const provider = this.providers.entries().next().value;
+          menuItem = this._setMenuItemAttributes(menuItem, provider[0], provider[1], downloadElement);
+          menuItem.setAttribute("label", "Move To " + provider[1].displayName);
+          menuItem.removeAttribute("hidden");
+        }
+      } else {
+        // Downloaded Item is saved in cloud storage provider folder
+
+        // Find local prefereed download directory path
+        // to display correct icon in Local Download context menu option
+        await CloudDownloadsInternal._setDefaultDownloadDirIconURL();
+
+        // Display context menus for download saved in cloud provider folder
+        if (this.providers.size > 1) {
+          let subMenuItem =  this._setMoveDownloadMenuPopUpAttributes(menuItem);
+          this.providers.forEach((value, key) => {
+            if (this._formatProviderName(value.displayName) !== downloadType) {
+              subMenuItem = this._setMenuItemAttributes(subMenuItem, key, value, downloadElement);
+              subMenuItem = subMenuItem.nextSibling;
+            }
+          });
+          subMenuItem = this._setMenuItemAttributes(subMenuItem, "local", {displayName: "Local Download"}, downloadElement);
+        } else {
+          menuItem = this._setMenuItemAttributes(menuItem, "local", {displayName: "Local Download"}, downloadElement);
+          menuItem.setAttribute("label", "Move To Local Download");
+          menuItem.removeAttribute("hidden");
+        }
+      }
+    }
+  },
+
+  handleEvent(event) {
+    // Handle multiple provider displayed as options in notification
+    if (event.target.parentElement.id === "multiProviderSelect") {
+      const cloudDownloadSave = event.currentTarget.children[1].children.cloudDownloadSave;
+      cloudDownloadSave.setAttribute("label", event.target.label);
+      cloudDownloadSave.setAttribute("providerKey", event.target.id);
+      cloudDownloadSave.removeAttribute("disabled");
+      return;
+    }
+
+    // Handle rendering right provider in context menu when shown
+    if (event.type === "contextmenu" && event.currentTarget.id === "downloadsListBox") {
+      const element = event.currentTarget.selectedItem;
+      if (!element) {
+        return;
+      }
+      this._displayMoveToCloudContextMenuItem(element);
```

Comment:
It's not really great that this is calling an async function, as it'll cause flickering as you update the UI after it shows (in fact, our menus are pretty fragile and I don't think that'll reliably work well across platforms).

Ideally, at the point where a context menu is shown you're able to provide the right information immediately. Can you cache the stuff you need here (looks like that's mostly an icon URI for the preferred download folder (which won't need to change unless the download folder changes) and checks related to the preferred download folder (ditto). It should be possible to just cache that information and update it when it changes. It looks like you could use the pref service to check for either of the 2 prefs changing, and you probably know best how to be notified if the cloud storage provider changes.

Reply:
Updated to cache icon URL and existing download settings check to remove await inside _displayMoveToCloudContextMenuItem function